cns/restserver/ipam.go

```diff
@@ -437,6 +438,65 @@ func (service *HTTPRestService) MarkIPAsPendingRelease(totalIpsToRelease int) (m
 	return pendingReleasedIps, nil
 }
 
+// MarkIPAsPendingRelease will attempt to set [n] number of ips to PendingRelease state.
+// It will start with any IPs in PendingProgramming state and then move on to any IPs in Allocated state
+// until it has reached the target release quantity.
+// If it is unable to set the expected number of IPs to PendingRelease, it will revert the changed IPs
+// and return an error.
+// MarkNIPsPendingRelease is no-op if [n] is not a positive integer.
+func (service *HTTPRestService) MarkNIPsPendingRelease(n int) (map[string]cns.IPConfigurationStatus, error) {
+	service.Lock()
+	defer service.Unlock()
+	// try to release from PendingProgramming
+	pendingProgrammingIPs := make(map[string]cns.IPConfigurationStatus)
+	for uuid, ipConfig := range service.PodIPConfigState { //nolint:gocritic // intentional value copy
+		if n <= 0 {
+			break
+		}
+		if ipConfig.GetState() == types.PendingProgramming {
+			updatedIPConfig, err := service.updateIPConfigState(uuid, types.PendingRelease, ipConfig.PodInfo)
+			if err != nil {
+				return nil, err
+			}
+
+			pendingProgrammingIPs[uuid] = updatedIPConfig
+			n--
+		}
+	}
+
+	// try to release from Available
+	availableIPs := make(map[string]cns.IPConfigurationStatus)
+	for uuid, ipConfig := range service.PodIPConfigState { //nolint:gocritic // intentional value copy
+		if n <= 0 {
+			break
+		}
+		if ipConfig.GetState() == types.Available {
+			updatedIPConfig, err := service.updateIPConfigState(uuid, types.PendingRelease, ipConfig.PodInfo)
+			if err != nil {
+				return nil, err
+			}
+
+			availableIPs[uuid] = updatedIPConfig
+			n--
+		}
+	}
+
+	// if we can release the requested quantity, return the IPs
+	if n <= 0 {
+		maps.Copy(pendingProgrammingIPs, availableIPs)
+		return pendingProgrammingIPs, nil
+	}
+
+	// else revert changes
+	for uuid, ipConfig := range pendingProgrammingIPs { //nolint:gocritic // intentional value copy
+		_, _ = service.updateIPConfigState(uuid, types.PendingProgramming, ipConfig.PodInfo)
```

Comment:
might be worth leaving a comment why this can't fail and why it's safe to ignore the returned `error`

Reply:
I'm not confident that it _is_ safe. But we don't really have a choice; If we get an error here it's because an IPID that we ***just changed while holding the state Mutex lock*** is no longer in our state. Maybe we should panic, because that probably indicates some cosmic ray/bit flip event has corrupted our internal state and we don't have any chance of resuming normal operation.